rbd/locks_test.go

```diff
@@ -0,0 +1,142 @@
+//go:build !nautilus && ceph_preview
+// +build !nautilus,ceph_preview
+
+package rbd
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestLocking(t *testing.T) {
+	conn := radosConnect(t)
+	require.NotNil(t, conn)
+	defer conn.Shutdown()
+
+	poolname := GetUUID()
+	err := conn.MakePool(poolname)
+	require.NoError(t, err)
+	defer conn.DeletePool(poolname)
+
+	ioctx, err := conn.OpenIOContext(poolname)
+	require.NoError(t, err)
+	defer ioctx.Destroy()
+
+	// Connect another cluster
+	conn1 := radosConnect(t)
+	require.NotNil(t, conn1)
+	defer conn1.Shutdown()
+
+	ioctx1, err := conn1.OpenIOContext(poolname)
+	require.NoError(t, err)
+	defer ioctx1.Destroy()
+
+	options := NewRbdImageOptions()
+	defer options.Destroy()
+	assert.NoError(t, options.SetUint64(ImageOptionOrder, uint64(testImageOrder)))
+
+	name := GetUUID()
+	err = CreateImage(ioctx, name, testImageSize, options)
+	assert.NoError(t, err)
+
+	t.Run("acquireLock", func(t *testing.T) {
+		img, err := OpenImage(ioctx, name, NoSnapshot)
+		assert.NoError(t, err)
+		defer func() {
+			assert.NoError(t, img.Close())
+		}()
+
+		// Shared lock mode is not supported
+		err = img.LockAcquire(LockModeShared)
+		assert.Error(t, err)
+
+		err = img.LockAcquire(LockModeExclusive)
+		assert.NoError(t, err)
+
+		isOwner, err := img.LockIsExclusiveOwner()
+		assert.NoError(t, err)
+		assert.True(t, isOwner)
+
+		err = img.LockBreak(LockModeExclusive, "not owner")
+		assert.Error(t, err)
+
+		err = img.LockRelease()
+		assert.NoError(t, err)
+	})
+
+	t.Run("listLock", func(t *testing.T) {
+		img, err := OpenImage(ioctx, name, NoSnapshot)
+		assert.NoError(t, err)
+		defer func() {
+			assert.NoError(t, img.Close())
+		}()
+
+		img1, err := OpenImage(ioctx1, name, NoSnapshot)
+		assert.NoError(t, err)
+		defer func() {
+			assert.NoError(t, img1.Close())
+		}()
+
+		err = img1.LockAcquire(LockModeExclusive)
+		assert.NoError(t, err)
+
+		err = img.LockAcquire(LockModeExclusive)
+		assert.Error(t, err)
+
+		isOwner, err := img.LockIsExclusiveOwner()
+		assert.NoError(t, err)
+		assert.False(t, isOwner)
+
+		locksList, err := img.LockGetOwners()
+		assert.NoError(t, err)
+		assert.Equal(t, len(locksList), 1)
+		assert.Equal(t, locksList[0].Mode, LockModeExclusive)
+
+		err = img.LockBreak(LockModeExclusive, locksList[0].Owner)
```

Comment:
Is it like a different owner can list the current lock owners of an image and break the other lock(using the owner key)? Sorry, I am thinking in pure POSIX advisory lock sematics where we have to wait for the other owner to release the lock and then new lock attempt is made.

Reply:
The test in ceph codebase I am referring to: https://github.com/ceph/ceph/blob/afb6b17aa77f7a85bf6d28b0174fa3c8609efc0d/src/test/librbd/test_librbd.cc#L10559-L10560

Reply:
+1 to the practice of comparing to tests in ceph. In this case, or in other cases where a reviewer asks "why" it can be helpful to put a comment in the code explaining that something is borrowed from the ceph code/tests/etc. A comment can be very helpful to future readers.

Reply:
Sure thing.
Added comments to the tests along with links that refer to the ceph codebase. PTAL